proposals/exception-handling/Exceptions-formal-overview.md

```diff
@@ -0,0 +1,207 @@
+
+```
+B^0 ::= val* [_] instr*
+B^k ::= catch_m{ tagaddr^? instr* }* B^k end | caught_m{ tagaddr val* } B^k end
+      | delegate{ labelidx } B^k end
+B^{k+1} ::= val* (label_n{instr*} B^k end) instr*
+```
+
+#### Throw Contexts
+
+Throw contexts don't skip over handlers, they are used to match a thrown exception with the innermost handler.
+
+```
+T ::= val* [_] instr* | label_n{instr*} T end | caught_m{ tagaddr val^n } T end
+   | frame_n{F} T end
+```
+
+Note that because `catch_n` instructions are not included above, there is always a unique maximal throw context.
+
+### Reduction of instructions
+
+Reduction steps for the new instructions or administrative instructions.
+
+```
+F; throw x  ↪  F; throw a  (if F.module.tagaddrs[x]=a)
+
+caught_m{a val^n} B^l[rethrow l] end
+  ↪ caught_m{a val^n} B^l[val^n (throw a)] end
+
+caught_m{a val^n} val^m end  ↪  val^m
+```
+
+An absent tag address in a `handler` clause (i.e., `a? = ε`) represents a `catch_all`.
+
+```
+F; val^n (try bt instr* (catch x instr'*)* (catch_all instr''*)? end)
+  ↪  F; label_m{} (catch_m{a instr'*}*{ε instr''*}? val^n instr* end) end
+  (if bt = [t1^n]→[t2^m] ∧ (F.module.tagaddrs[x]=a)*)
+
+catch_m{a? instr*}* val^m end ↪ val^m
+
+S; F; catch_m{a1? instr*}{a'? instr'*}* T[val^n (throw a)] end
+  ↪  S; F; caught_m{a val^n} (val^n)? instr* end
+  (if (a1? = ε ∨ a1? = a) ∧ S.tags(a).type = [t^n]→[])
+
+S; F; catch_m{a1? instr*}{a'? instr'*}* T[val^n (throw a)] end
+  ↪  S; F; catch_m{a'? instr'*}* T[val^n (throw a)] end
+  (if a1? ≠ ε ∧ a1? ≠ a)
+
+S; F; catch_m T[val^n (throw a)] end
+  ↪  S; F; val^n (throw a)
+
+
+val^n (try bt instr* delegate l)
+  ↪ label_m{} (delegate{l} val^n instr* end) end
+  (if bt = [t^n]→[t^m])
+
+delegate{l} val^n end ↪ val^n
+
+label_m{} B^l[ delegate{l} T[val^n (throw a)] end ] end
+  ↪ val^n (throw a)
+```
+
+Note that the last reduction step above is similar to the reduction of `br`.
+
+### Typing rules for administrative instructions
+
+```
+S.tags[a].type = [t*]→[]
+--------------------------------
+S;C ⊢ throw a : [t1* t*]→[t2*]
+
+((S.tags[a].type = [t'*]→[])?
+ S;C, labels {result [t*], kind catch} ⊢ instr'* : [t'*?]→[t*])*
+S;C, labels [t*] ⊢ instr* : []→[t*]
+-----------------------------------------------------------------
+S;C, labels [t*] ⊢ catch_n{a? instr'*}* instr* end : []→[t*]
+
+S;C, labels [t*] ⊢ instr* : []→[t*]
+C.labels[l] = [t'*]
+------------------------------------------------------
+S;C, labels [t*] ⊢ delegate{l} instr* end : []→[t*]
+
+S.tags[a].type = [t'*]→[]
+(val:t')*
+S;C, labels {result [t*], kind catch} ⊢ instr* : []→[t*]
+----------------------------------------------------------
+S;C, labels [t*] ⊢ caught_m{a val^n} instr* end : []→[t*]
+```
+
```

Comment:
The first one should already be covered by the definition of `T`, which  includes `frame`.

The second is a bit more subtle. We'll need to extend the definition of "terminal", as given in the [Appendix](https://webassembly.github.io/spec/core/appendix/properties.html#theorems) – this has to allow `val* T[throw e] instr*` as a new form of terminal thread configuration, which would be the representation of a computation crashed with an uncaught exception. That does not actually require any additional reduction rule.

(One could try to reduce this to a simpler form like `val* (throw e)`, but that's somewhat tricky, since you'd need a rule that reduces the former to the latter but deterministically only applies at the outermost level. That cannot really be expressed within the current set-up, I believe.)

Reply:
I think addressing this as a followup is fine.

Reply:
Added a couple of lines in the end, mentioning that we haven't addressed uncaught exceptions yet, as a TODO for a follow up. Resolving this discussion for now.